### PR TITLE
[FEATURE] modularisation refining breakends

### LINKED ISSUE #42: [FEATURE] Modularize methods for refining breakends

### PR DESCRIPTION
**Pull request by Irallia**
Resolves #42 .

I added the new parser option -r with 2 different refining methods and the possibility to not use one. (none is implemented)

### CODE REVIEW

File: src/detect_breakends/junction_detection.cpp [hunk 1]
```diff
@@ -365,6 +369,20 @@ void detect_junctions_in_alignment_file(const std::filesystem::path & alignment_
             break;
     }
 
+    switch (refinement_method)
+    {
+        case 0: // no refinement
+            seqan3::debug_stream << "No refinement was selected.\n";
+            break;
+        case 1: // sViper
+            seqan3::debug_stream << "The sViper refinement method is not jet implemented\n";
```
Comment (eaasna) [inline src/detect_breakends/junction_detection.cpp:378]:
```suggestion
            seqan3::debug_stream << "The sViper refinement method is not yet implemented\n";
```

File: src/detect_breakends/junction_detection.cpp [hunk 2]
```diff
@@ -365,6 +369,20 @@ void detect_junctions_in_alignment_file(const std::filesystem::path & alignment_
...
+            break;
+        case 2: // prioritzation_based_on_rank_aggregation
+            seqan3::debug_stream << "The prioritzation_based_on_rank_aggregation refinement method is not jet"
```
Comment (eaasna) [inline src/detect_breakends/junction_detection.cpp:381]:
```suggestion
            seqan3::debug_stream << "The prioritzation_based_on_rank_aggregation refinement method is not yet"
```

File: include/detect_breakends/junction_detection.hpp [hunk 3]
```diff
@@ -21,6 +21,9 @@
  *                                                           1: hierarchical_clustering,
  *                                                           2: self-balancing_binary_tree,
  *                                                           3: candidate_selection_based_on_voting)
+ * \param refinement_method method for refining breakends (0: no_refinement,
+ *                                                         1: sViper,
+ *                                                         2: prioritzation_based_on_rank_aggregation)
  *
  * \details Detects junctions from the CIGAR strings and supplementary alignment tags of read alignment records.
```
Comment (eaasna) [inline include/detect_breakends/junction_detection.hpp:28]: I think that junctions and breakends represent places where the query sequence differs from the reference. But I'm not quite sure about the exact meaning of the two terms

**Review by eaasna — COMMENTED** [review]
Nice work :) 

just a couple of typos (and a question below)

File: src/detect_breakends/junction_detection.cpp (hunk 1 above)
Comment (Irallia) [reply src/detect_breakends/junction_detection.cpp:378]: and I did it again ^^ thanks for the hint 😅

File: include/detect_breakends/junction_detection.hpp (hunk 3 above)
Comment (eaasna) [reply include/detect_breakends/junction_detection.hpp:28]: Thanks for the explanation! :)

**Review by eaasna — APPROVED** [review]

File: include/detect_breakends/junction_detection.hpp [hunk 4]
```diff
@@ -22,6 +22,9 @@
  *                                                           1: hierarchical_clustering,
  *                                                           2: self-balancing_binary_tree,
  *                                                           3: candidate_selection_based_on_voting)
+ * \param refinement_method method for refining breakends (0: no_refinement,
+ *                                                         1: sViper,
+ *                                                         2: prioritzation_based_on_rank_aggregation)
```
Comment (joshuak94) [inline include/detect_breakends/junction_detection.hpp:27]: Prioritization spelling

Comment (joshuak94) [reply include/detect_breakends/junction_detection.hpp:27]: I'm not quite sure what you mean with the prioritization based on rank aggregation part. If this adapted from Vaquita, then it doesn't belong in refining breakpoints.

The prioritization used in Vaquita is more to "rank" the variants by how important they are. This also has the effect that the variants which are considered "more important" are more likely to be true variants. However, rank aggregation does not do anything to the existing breakpoints. It's sort of a "final step" in the pipeline.

For now, I would think that the only breakend refinement we have is via sViper.

File: src/detect_breakends.cpp [hunk 5]
```diff
@@ -23,14 +23,32 @@ auto enumeration_names(clustering_methods)
...
+                                                   refinement_methods::prioritzation_based_on_rank_aggregation}};
+};
 
 struct cmd_arguments
 {
     std::filesystem::path alignment_file_path{};
     std::filesystem::path insertion_file_path{};
     std::vector<uint8_t> methods{1, 2, 3, 4};                   // default is using all methods
     clustering_methods clustering_method{simple_clustering};    // default is the simple clustering method
+    refinement_methods refinement_method{no_refinement};        // default is using no refinement
```
Comment (joshuak94) [inline src/detect_breakends.cpp:51]: Why not make the default to using sViper?

Comment (joshuak94) [reply src/detect_breakends.cpp:51]: Also we have to consider that sViper will only work if the user provides long reads AND short reads. With only one dataset given, the command parser should probably throw an error if sViper refinement is selected.

**Review by joshuak94 — CHANGES_REQUESTED** [review]
Looks good! But to be honest, I don't know why we want to give this as an option. I think it would be fine to have this be automatically handled behind the scenes. AKA if only short or only long reads are given, then sViper is not run. However, if both are given, sViper should be run automatically.

Also the rank aggregation doesn't belong with breakpoint refinement, I think. It's just calculating a score basically based on what evident is present and how strongly it's present. It doesn't actually modify any of the variants.

File: include/detect_breakends/junction_detection.hpp (hunk 4 above)
Comment (Irallia) [reply include/detect_breakends/junction_detection.hpp:27]: Yes you are right, the `prioritzation_based_on_rank_aggregation` is probably wrong here. I spoke to david and we have still 2 different refining methods. The other is a method from SVirl.
On the other side, I think we want to able and enable the refinement, as it also depends on the input. sViper only works if we have short and long reads. The method from sVirl works with long reads, and we have none for short reads only.

We could remove this parameter later and could do this decision automaticaly but for now it is easyer for us and we are able to add more if there are more methods.

File: src/detect_breakends.cpp (hunk 5 above)
Comment (Irallia) [reply src/detect_breakends.cpp:51]: Yes, we have to solve https://github.com/seqan/igenvar/issues/43 next, and there we could add a check if sViper is possible to use when we know which input we have.

File: include/method_enums.hpp [hunk 6]
```diff
@@ -8,3 +8,11 @@ enum clustering_methods
     self_balancing_binary_tree = 2,
     candidate_selection_based_on_voting = 3
 };
+
+//!\brief An enum for the different refinement methods.
+enum refinement_methods
+{
+    no_refinement = 0,
+    sViper = 1,
+    sVirl_refinement_method = 2
```
Comment (joshuak94) [inline include/method_enums.hpp:17]: Small thing, we should say `sViper_refinement_method` to keep it consistent with the naming.

File: src/detect_breakends/junction_detection.cpp [hunk 7]
```diff
@@ -373,6 +377,20 @@ void detect_junctions_in_alignment_file(const std::filesystem::path & alignment_
...
+    switch (refinement_method)
+    {
+        case 0: // no refinement
+            seqan3::debug_stream << "No refinement was selected.\n";
+            break;
+        case 1: // sViper
+            seqan3::debug_stream << "The sViper refinement method is not yet implemented\n";
+            break;
+        case 2: // prioritzation_based_on_rank_aggregation
+            seqan3::debug_stream << "The prioritzation_based_on_rank_aggregation refinement method is not yet"
```
Comment (joshuak94) [inline src/detect_breakends/junction_detection.cpp:389]:
```suggestion
            seqan3::debug_stream << "The sVirl refinement method is not yet"
```

**Review by joshuak94 — CHANGES_REQUESTED** [review]
I think there should also be tests to cover the sViper and sVirl refinement cases, even if they aren't yet implemented. Just so that in the future, any changes will be covered by the tests already.

**Comment by Irallia** [comment]
You are right about the missing tests. I will add in an extra PR the others for the other parameters. I hope these are the tests you meant.

File: test/api/junction_detection_test.cpp [hunk 8]
```diff
@@ -120,6 +125,65 @@ TEST(junction_detection, method_1_and_2)
...
+    testing::internal::CaptureStdout();
+    detect_junctions_in_alignment_file(DATADIR"simulated.minimap2.hg19.coordsorted_cutoff.sam",
+                                       tmp_dir/"detect_breakends_out_short.fasta",
+                                       {1, 2, 3, 4},
+                                       simple_clustering,
+                                       sViper_refinement_method,
+                                       sv_default_length);
+
+    std::string std_cout = testing::internal::GetCapturedStdout();
+    seqan3::debug_stream << "std_out:\n" << std_cout << '\n';
```
Comment (joshuak94) [inline test/api/junction_detection_test.cpp:162]: Is there a reason you decided to print the `std::out` output, but not actually test that it gives what's expected? Just wondering. I would say we should either test the output, or just remove the printing.

**Review by joshuak94 — APPROVED** [review]

File: test/api/junction_detection_test.cpp (hunk 8 above)
Comment (Irallia) [reply test/api/junction_detection_test.cpp:162]: Doesn't the line after that test just that?

**Pull request merged**